File: 2018/day9/main.py
#!/usr/bin/env python

import os
import sys
from collections import defaultdict

sys.path.insert(0, os.path.abspath('../..'))

from utils import solve  # nopep8


class Circle:
    placed = []
    left = set()
    current = None

    def __init__(self, last):

        self.placed = [0]
        self.current = 0
        self.left = set(i for i in range(1, last + 1))

    def place(self):
        m = min(self.left)
        if m % 23 == 0:
            return self.place23(m)

        if len(self.placed) in [1, 2]:
            l = 1
        elif self.current == len(self.placed) - 1:
            l = 1
        else:
            l = (self.current + 2) % (len(self.placed) + 1)
        self.placed.insert(l, m)
        self.left.remove(m)
        self.current = l
        return 0

    def place23(self, m):
        l = (self.current - 7) % len(self.placed)
        s = m + self.placed.pop(l)
        self.left.remove(m)
        self.current = l
        return s


def p1(lines):
    p, last = lines
    scores = defaultdict(int)
    current_p = 0
    circle = Circle(last)
    while len(circle.left) > 0:
        score = circle.place()
        scores[current_p] += score
        current_p = (current_p + 1) % p

    return max(scores.values())


def p2(_):
    return None


if __name__ == "__main__":
    solve((424, 71482), lambda x: x, p1, p2)

File: 2018/day9/main.go
package main

import (
	"container/list"
	"fmt"
)

type Circle struct {
	placed *list.List
	marbles int
	currentMarble int
	currentPos *list.Element
}

func (c *Circle) Completed() bool {
	return c.currentMarble == c.marbles
}

func (c *Circle) Place() int {
	if c.currentMarble % 23 == 0 {
		c.Prev().Prev().Prev().Prev().Prev().Prev().Prev()
		toRemove := c.currentPos
		s := toRemove.Value.(int) + c.currentMarble
		c.Next()
		c.placed.Remove(toRemove)
		c.currentMarble += 1
		return s
	} else {
		c.Next()
		c.currentPos = c.placed.InsertAfter(c.currentMarble, c.currentPos)
		c.currentMarble += 1
		return 0
	}
}

func (c *Circle) Next() *Circle {
	if e := c.currentPos.Next(); e != nil {
		c.currentPos = e
	} else {
		c.currentPos = c.placed.Front()
	}
	return c
}

func (c *Circle) Prev() *Circle {
	if e := c.currentPos.Prev(); e != nil {
		c.currentPos = e
	} else {
		c.currentPos = c.placed.Back()
	}
	return c
}

func (c *Circle) Print() {
	fmt.Print("Circle: ")
	for el := c.placed.Front(); el != nil; el = el.Next() {
		fmt.Printf("%d ", el.Value)
	}
	fmt.Print("\n")
}

func NewCircle(last int) *Circle {
	c := Circle{
		placed: list.New(),
		marbles: last,
		currentMarble: 1,
	}
	c.placed.PushFront(0)
	c.currentPos = c.placed.Front()
	return &c
}

func main() {
	solve(424, 71482)
	solve(424, 71482*100)
}

func solve(players, last int) {
	scores := map[int]int{}
	circle := NewCircle(last)
	for currentPlayer := 0;!circle.Completed(); {
		s:= circle.Place()
		if s > 0 {
			scores[currentPlayer] += s
		}
		currentPlayer = (currentPlayer + 1) % players
	}
	max := 0
	for _, s := range scores {
		if s > max {
			max = s
		}
	}
	println(max)
}
